{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `pymatgen.io.vasp.help` 模块使用"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymatgen.io.vasp.help import VaspDoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/leubuntu/src/miniconda3/envs/base_ysl/lib/python3.11/site-packages/urllib3/connectionpool.py:1103: InsecureRequestWarning: Unverified HTTPS request is being made to host '127.0.0.1'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings\n",
      "  warnings.warn(\n",
      "/home/leubuntu/src/miniconda3/envs/base_ysl/lib/python3.11/site-packages/pymatgen/io/vasp/help.py:70: GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system (\"lxml\"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.\n",
      "\n",
      "The code that caused this warning is on line 70 of the file /home/leubuntu/src/miniconda3/envs/base_ysl/lib/python3.11/site-packages/pymatgen/io/vasp/help.py. To get rid of this warning, pass the additional argument 'features=\"lxml\"' to the BeautifulSoup constructor.\n",
      "\n",
      "  soup = BeautifulSoup(response.text)\n",
      "/home/leubuntu/src/miniconda3/envs/base_ysl/lib/python3.11/site-packages/urllib3/connectionpool.py:1103: InsecureRequestWarning: Unverified HTTPS request is being made to host '127.0.0.1'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['ADDGRID',\n",
       " 'AEXX',\n",
       " 'AGGAC',\n",
       " 'AGGAX',\n",
       " 'ALDAC',\n",
       " 'ALDAX',\n",
       " 'ALGO',\n",
       " 'ALPHA VDW',\n",
       " 'AMGGAC',\n",
       " 'AMGGAX',\n",
       " 'AMIN',\n",
       " 'AMIX',\n",
       " 'AMIX MAG',\n",
       " 'ANDERSEN PROB',\n",
       " 'ANTIRES',\n",
       " 'APACO',\n",
       " 'BANDGAP',\n",
       " 'BEXT',\n",
       " 'BMIX',\n",
       " 'BMIX MAG',\n",
       " 'BPARAM',\n",
       " 'BSEELECTRON',\n",
       " 'BSEHOLE',\n",
       " 'BSEPREC',\n",
       " 'CH LSPEC',\n",
       " 'CH NEDOS',\n",
       " 'CH SIGMA',\n",
       " 'CLL',\n",
       " 'CLN',\n",
       " 'CLNT',\n",
       " 'CLZ',\n",
       " 'CMBJ',\n",
       " 'CMBJA',\n",
       " 'CMBJB',\n",
       " 'CMBJE',\n",
       " 'CPARAM',\n",
       " 'CSHIFT',\n",
       " 'CUTOFF MU',\n",
       " 'CUTOFF SIGMA',\n",
       " 'CUTOFF TYPE',\n",
       " 'DEPER',\n",
       " 'DIMER DIST',\n",
       " 'DIPOL',\n",
       " 'DQ',\n",
       " 'EBREAK',\n",
       " 'EDIFF',\n",
       " 'EDIFFG',\n",
       " 'EFERMI',\n",
       " 'EFIELD',\n",
       " 'EFIELD PEAD',\n",
       " 'EINT',\n",
       " 'EMAX',\n",
       " 'EMIN',\n",
       " 'ENAUG',\n",
       " 'ENCUT',\n",
       " 'ENCUTFOCK',\n",
       " 'ENCUTGW',\n",
       " 'ENCUTGWSOFT',\n",
       " 'ENINI',\n",
       " 'EPSILON',\n",
       " 'ESTOP',\n",
       " 'EVENONLY',\n",
       " 'EVENONLYGW',\n",
       " 'FERDO',\n",
       " 'FERWE',\n",
       " 'FINDIFF',\n",
       " 'FOCKCORR',\n",
       " 'GAMMA VDW',\n",
       " 'GGA',\n",
       " 'GGA COMPAT',\n",
       " 'HFALPHA',\n",
       " 'HFLMAX',\n",
       " 'HFRCUT',\n",
       " 'HFSCREEN',\n",
       " 'HILLS BIN',\n",
       " 'HILLS H',\n",
       " 'HILLS W',\n",
       " 'HITOLER',\n",
       " 'I CONSTRAINED M',\n",
       " 'IALGO',\n",
       " 'IALL IN ONE',\n",
       " 'IBRION',\n",
       " 'ICHARG',\n",
       " 'ICHIBARE',\n",
       " 'ICORELEVEL',\n",
       " 'IDIPOL',\n",
       " 'IEPSILON',\n",
       " 'IGPAR',\n",
       " 'IMAGES',\n",
       " 'IMIX',\n",
       " 'INCREM',\n",
       " 'INIMIX',\n",
       " 'INIWAV',\n",
       " 'IPEAD',\n",
       " 'IRC DELTA0',\n",
       " 'IRC DIRECTION',\n",
       " 'IRC MAXSTEP',\n",
       " 'IRC MINSTEP',\n",
       " 'IRC STOP',\n",
       " 'IRC VNORM0',\n",
       " 'ISIF',\n",
       " 'ISMEAR',\n",
       " 'ISPIN',\n",
       " 'ISTART',\n",
       " 'ISYM',\n",
       " 'IVDW',\n",
       " 'IVDW NL',\n",
       " 'IWAVPR',\n",
       " 'KBLOCK',\n",
       " 'KGAMMA',\n",
       " 'KPAR',\n",
       " 'KPOINT BSE',\n",
       " 'KPOINTS OPT NKBATCH',\n",
       " 'KSPACING',\n",
       " 'LADDER',\n",
       " 'LAECHG',\n",
       " 'LALL IN ONE',\n",
       " 'LAMBDA',\n",
       " 'LANGEVIN GAMMA',\n",
       " 'LANGEVIN GAMMA L',\n",
       " 'LASPH',\n",
       " 'LASYNC',\n",
       " 'LATTICE CONSTRAINTS',\n",
       " 'LBERRY',\n",
       " 'LBLUEOUT',\n",
       " 'LBONE',\n",
       " 'LCALCEPS',\n",
       " 'LCALCPOL',\n",
       " 'LCHARG',\n",
       " 'LCHARGH5',\n",
       " 'LCHIMAG',\n",
       " 'LCORR',\n",
       " 'LDAU',\n",
       " 'LDAUJ',\n",
       " 'LDAUL',\n",
       " 'LDAUPRINT',\n",
       " 'LDAUTYPE',\n",
       " 'LDAUU',\n",
       " 'LDIAG',\n",
       " 'LDIPOL',\n",
       " 'LDISENTANGLED',\n",
       " 'LEFG',\n",
       " 'LELF',\n",
       " 'LEPSILON',\n",
       " 'LFINITE TEMPERATURE',\n",
       " 'LFOCKACE',\n",
       " 'LFOCKAEDFT',\n",
       " 'LFXC',\n",
       " 'LH5',\n",
       " 'LHARTREE',\n",
       " 'LHFCALC',\n",
       " 'LHYPERFINE',\n",
       " 'LIBMBD ALPHA',\n",
       " 'LIBMBD C6AU',\n",
       " 'LIBMBD K GRID',\n",
       " 'LIBMBD K GRID SHIFT',\n",
       " 'LIBMBD MBD A',\n",
       " 'LIBMBD MBD BETA',\n",
       " 'LIBMBD METHOD',\n",
       " 'LIBMBD N OMEGA GRID',\n",
       " 'LIBMBD PARALLEL MODE',\n",
       " 'LIBMBD R0AU',\n",
       " 'LIBMBD TS D',\n",
       " 'LIBMBD TS SR',\n",
       " 'LIBMBD VDW PARAMS KIND',\n",
       " 'LIBMBD XC',\n",
       " 'LIBXC1',\n",
       " 'LIBXC1 Pn',\n",
       " 'LIBXC2',\n",
       " 'LIBXC2 Pn',\n",
       " 'LKPOINTS OPT',\n",
       " 'LKPOINTS WAN',\n",
       " 'LKPROJ',\n",
       " 'LLRAUG',\n",
       " 'LMAXFOCK',\n",
       " 'LMAXFOCKAE',\n",
       " 'LMAXMIX',\n",
       " 'LMAXPAW',\n",
       " 'LMAXTAU',\n",
       " 'LMIXTAU',\n",
       " 'LMODELHF',\n",
       " 'LMONO',\n",
       " 'LMP2LT',\n",
       " 'LNABLA',\n",
       " 'LNMR SYM RED',\n",
       " 'LNONCOLLINEAR',\n",
       " 'LOCPROJ',\n",
       " 'LOPTICS',\n",
       " 'LORBIT',\n",
       " 'LORBMOM',\n",
       " 'LPARD',\n",
       " 'LPEAD',\n",
       " 'LPHON DISPERSION',\n",
       " 'LPHON POLAR',\n",
       " 'LPHON READ FORCE CONSTANTS',\n",
       " 'LPLANE',\n",
       " 'LREAL',\n",
       " 'LRHFCALC',\n",
       " 'LRPA',\n",
       " 'LRPAFORCE',\n",
       " 'ML IAFILT2',\n",
       " 'ML IALGO LINREG',\n",
       " 'ML ICOUPLE',\n",
       " 'ML ICRITERIA',\n",
       " 'ML IERR',\n",
       " 'ML IREG',\n",
       " 'ML ISCALE TOTEN',\n",
       " 'ML ISTART',\n",
       " 'ML IWEIGHT',\n",
       " 'ML LAFILT2',\n",
       " 'ML LBASIS DISCARD',\n",
       " 'ML LCOUPLE',\n",
       " 'ML LEATOM',\n",
       " 'ML LERR',\n",
       " 'ML LFAST',\n",
       " 'ML LHEAT',\n",
       " 'ML LMAX2',\n",
       " 'ML LMLFF',\n",
       " 'ML LSPARSDES',\n",
       " 'ML MB',\n",
       " 'ML MB MIN',\n",
       " 'ML MCONF',\n",
       " 'ML MCONF NEW',\n",
       " 'ML MHIS',\n",
       " 'ML MODE',\n",
       " 'ML MRB1',\n",
       " 'ML MRB2',\n",
       " 'ML NATOM COUPLED',\n",
       " 'ML NHYP',\n",
       " 'ML NMDINT',\n",
       " 'ML NRANK SPARSDES',\n",
       " 'ML OUTBLOCK',\n",
       " 'ML OUTPUT MODE',\n",
       " 'ML RCOUPLE',\n",
       " 'ML RCUT1',\n",
       " 'ML RCUT2',\n",
       " 'ML RDES SPARSDES',\n",
       " 'ML SCLC CTIFOR',\n",
       " 'ML SIGV0',\n",
       " 'ML SIGW0',\n",
       " 'ML SION1',\n",
       " 'ML SION2',\n",
       " 'ML W1',\n",
       " 'ML WTIFOR',\n",
       " 'ML WTOTEN',\n",
       " 'ML WTSIF',\n",
       " 'NATURALO',\n",
       " 'NBANDS',\n",
       " 'NBANDS WAVE',\n",
       " 'NBANDSEXACT',\n",
       " 'NBANDSGW',\n",
       " 'NBANDSO',\n",
       " 'NBANDSV',\n",
       " 'NBLK',\n",
       " 'NBLOCK',\n",
       " 'NBMOD',\n",
       " 'NBSEEIG',\n",
       " 'NCORE',\n",
       " 'NCORE IN IMAGE1',\n",
       " 'NCRPA BANDS',\n",
       " 'NEDOS',\n",
       " 'NELECT',\n",
       " 'NELM',\n",
       " 'NELMDL',\n",
       " 'NELMGW',\n",
       " 'NELMIN',\n",
       " 'NFREE',\n",
       " 'NGX',\n",
       " 'NGXF',\n",
       " 'NGY',\n",
       " 'NGYF',\n",
       " 'NGYROMAG',\n",
       " 'NGZ',\n",
       " 'NGZF',\n",
       " 'NHC NCHAINS',\n",
       " 'NHC NRESPA',\n",
       " 'NHC NS',\n",
       " 'NKRED',\n",
       " 'NKREDX',\n",
       " 'NKREDY',\n",
       " 'NKREDZ',\n",
       " 'NLSPLINE',\n",
       " 'NMAXFOCKAE',\n",
       " 'NMAXFOCKAE and LMAXFOCKAE',\n",
       " 'NOMEGA',\n",
       " 'NOMEGA DUMP',\n",
       " 'NOMEGAPAR',\n",
       " 'NOMEGAR',\n",
       " 'NPACO',\n",
       " 'NPAR',\n",
       " 'NPPSTR',\n",
       " 'NRMM',\n",
       " 'NSIM',\n",
       " 'NSTORB',\n",
       " 'NSUBSYS',\n",
       " 'NSW',\n",
       " 'NTARGET STATES',\n",
       " 'NTAUPAR',\n",
       " 'NTEMPER',\n",
       " 'NUM WANN',\n",
       " 'NUPDOWN',\n",
       " 'NWRITE',\n",
       " 'ODDONLY',\n",
       " 'ODDONLYGW',\n",
       " 'OFIELD A',\n",
       " 'OFIELD KAPPA',\n",
       " 'OFIELD Q6 FAR',\n",
       " 'OFIELD Q6 NEAR',\n",
       " 'OMEGAMAX',\n",
       " 'OMEGAMIN',\n",
       " 'OMEGATL',\n",
       " 'PARAM1',\n",
       " 'PARAM2',\n",
       " 'PFLAT',\n",
       " 'PHON BORN CHARGES',\n",
       " 'PHON DIELECTRIC',\n",
       " 'PHON DOS',\n",
       " 'PHON G CUTOFF',\n",
       " 'PHON LBOSE',\n",
       " 'PHON LMC',\n",
       " 'PHON NEDOS',\n",
       " 'PHON NSTRUCT',\n",
       " 'PHON NTLIST',\n",
       " 'PHON NWRITE',\n",
       " 'PHON SIGMA',\n",
       " 'PHON TLIST',\n",
       " 'PLEVEL',\n",
       " 'PMASS',\n",
       " 'POMASS',\n",
       " 'POTIM',\n",
       " 'PREC',\n",
       " 'PRECFOCK',\n",
       " 'Profiling',\n",
       " 'PROUTINE',\n",
       " 'PSTRESS',\n",
       " 'PSUBSYS',\n",
       " 'PTHRESHOLD',\n",
       " 'QMAXFOCKAE',\n",
       " 'QSPIRAL',\n",
       " 'QUAD EFG',\n",
       " 'RANDOM SEED',\n",
       " 'ROPT',\n",
       " 'RSMBJ',\n",
       " 'RWIGS',\n",
       " 'SAXIS',\n",
       " 'SCALEE',\n",
       " 'SCSRAD',\n",
       " 'SHAKEMAXITER',\n",
       " 'SHAKETOL',\n",
       " 'SIGMA',\n",
       " 'SMASS',\n",
       " 'SMBJ',\n",
       " 'SMEARINGS',\n",
       " 'SPRING',\n",
       " 'STEP MAX',\n",
       " 'STEP SIZE',\n",
       " 'STOP ON',\n",
       " 'SYMPREC',\n",
       " 'SYSTEM',\n",
       " 'TEBEG',\n",
       " 'TEEND',\n",
       " 'TIME',\n",
       " 'TSUBSYS',\n",
       " 'VACPOTFLAT',\n",
       " 'VALUE MAX',\n",
       " 'VALUE MIN',\n",
       " 'VCA',\n",
       " 'VCAIMAGES',\n",
       " 'VCUTOFF',\n",
       " 'VDW A1',\n",
       " 'VDW A2',\n",
       " 'VDW ALPHA',\n",
       " 'VDW C6',\n",
       " 'VDW C6AU',\n",
       " 'VDW CNRADIUS',\n",
       " 'VDW D',\n",
       " 'VDW R0',\n",
       " 'VDW R0AU',\n",
       " 'VDW RADIUS',\n",
       " 'VDW S6',\n",
       " 'VDW S8',\n",
       " 'VDW SR',\n",
       " 'VOSKOWN',\n",
       " 'WANNIER90 WIN',\n",
       " 'WC',\n",
       " 'WEIMIN',\n",
       " 'WRT POTENTIAL',\n",
       " 'XC',\n",
       " 'XC C',\n",
       " 'ZAB VDW',\n",
       " 'ZVAL']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 静态方法\n",
    "VaspDoc.get_incar_tags()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/leubuntu/src/miniconda3/envs/base_ysl/lib/python3.11/site-packages/urllib3/connectionpool.py:1103: InsecureRequestWarning: Unverified HTTPS request is being made to host '127.0.0.1'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings\n",
      "  warnings.warn(\n",
      "/home/leubuntu/src/miniconda3/envs/base_ysl/lib/python3.11/site-packages/pymatgen/io/vasp/help.py:51: GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system (\"lxml\"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.\n",
      "\n",
      "The code that caused this warning is on line 51 of the file /home/leubuntu/src/miniconda3/envs/base_ysl/lib/python3.11/site-packages/pymatgen/io/vasp/help.py. To get rid of this warning, pass the additional argument 'features=\"lxml\"' to the BeautifulSoup constructor.\n",
      "\n",
      "  soup = BeautifulSoup(response.text)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'IBRION\\xa0= -1 | 0 | 1 | 2 | 3 | 5 | 6 | 7 | 8 | 40 | 44\\xa0\\n\\nDefault: IBRION\\n\\n= -1\\n\\nfor NSW=−1 or 0\\n\\n= 0\\n\\nelse\\n\\nDescription: IBRION determines how the ions are updated and moved.\\n\\nFor IBRION=0, a molecular dynamics is performed, whereas all other algorithms are destined for relaxations into a local energy minimum. For difficult relaxation problems it is recommended to use the conjugate gradient algorithm (IBRION=2), which presently possesses the most reliable backup routines. Damped molecular dynamics (IBRION=3) are often useful when starting from very bad initial guesses. Close to the local minimum the RMM-DIIS (IBRION=1) is usually the best choice. IBRION=5 and IBRION=6 are using finite differences to determine the second derivatives (Hessian matrix and phonon frequencies), whereas IBRION=7 and IBRION=8 use density functional perturbation theory to calculate the derivatives.\\n\\nIBRION=-1: no update.\\nThe ions are not moved, but NSW outer loops are performed. In each outer loop the electronic degrees of freedom are re-optimized (for NSW>0 this obviously does not make much sense, except for test purposes). If no ionic update is required use NSW=0 instead.\\n\\nIBRION=0: molecular dynamics.\\nStandard ab-initio molecular dynamics. A Verlet algorithm (or fourth-order predictor-corrector if VASP was linked with stepprecor.o) is used to integrate Newton\\'s equations of motion. POTIM supplies the timestep in femto seconds. The parameter SMASS provides additional control.\\n\\nTip: NWRITE=0,1\\n\\nIBRION=1: ionic relaxation (RMM-DIIS).\\nFor IBRION=1, a quasi-Newton (variable metric) algorithm is used to relax the ions into their instantaneous groundstate. The forces and the stress tensor are used to determine the search directions for finding the equilibrium positions (the total energy is not taken into account). This algorithm is very fast and efficient close to local minima, but fails badly if the initial positions are a bad guess (use IBRION=2 in that case). Since the algorithm builds up an approximation of the Hessian matrix it requires very accurate forces, otherwise it will fail to converge. An efficient way to achieve this is to set NELMIN to a value between 4 and 8 (for simple bulk materials 4 is usually adequate, whereas 8 might be required for complex surfaces where the charge density converges very slowly). This forces a minimum of 4 to 8 electronic steps between each ionic step, and guarantees that the forces are well converged at each step.\\nThe implemented algorithm is called RMM-DIIS.[1] It implicitly calculates an approximation of the inverse Hessian matrix by taking into account information from previous iterations. On startup, the initial Hessian matrix is diagonal and equal to POTIM. Information from old steps (which can lead to linear dependencies) is automatically removed from the iteration history, if required. The number of vectors kept in the iterations history (which corresponds to the rank of the Hessian matrix must not exceed the degrees of freedom. Naively the number of degrees of freedom is 3(NIONS-1). But symmetry arguments or constraints can reduce this number significantly.\\nThere are two algorithms build in to remove information from the iteration history:\\n\\nIf NFREE is set in the INCAR file, only up to NFREE ionic steps are kept in the iteration history (the rank of the approximate Hessian matrix is not larger than NFREE).\\nIf NFREE is not specified, the criterion whether information is removed from the iteration history is based on the eigenvalue spectrum of the inverse Hessian matrix: if one eigenvalue of the inverse Hessian matrix is larger than 8, information from previous steps is discarded.\\nFor complex problems NFREE can usually be set to a rather large value (i.e. 10-20), however systems of low dimensionality require a carful setting of NFREE (or preferably an exact counting of the number of degrees of freedom). To increase NFREE beyond 20 rarely improves convergence. If NFREE is set to too large, the RMM-DIIS algorithm might diverge.\\nThe choice of a reasonable POTIM is also important and can speed up calculations significantly, we recommend to find an optimal POTIM using IBRION=2 or performing a few test calculations (see below).\\n\\nIBRION=2: ionic relaxation (conjugate gradient algorithm).\\nA conjugate-gradient algorithm (a discussion of this algorithm can be found for instance in Numerical Recipes, by Press et al.[2]) is used to relax the ions into their instantaneous groundstate. In the first step ions (and cell shape) are changed along the direction of the steepest descent (i.e. the direction of the calculated forces and stress tensor). The conjugate gradient method requires a line minimization, which is performed in several steps: \\n\\nFirst a trial step into the search direction (scaled gradients) is done, with the length of the trial step controlled by the POTIM tag. Then the energy and the forces are recalculated.\\nThe approximate minimum of the total energy is calculated from a cubic (or quadratic) interpolation taking into account the change of the total energy and the change of the forces (3 pieces of information), then a corrector step to the approximate minimum is performed.\\nAfter the corrector step the forces and energy are recalculated and it is checked whether the forces contain a significant component parallel to the previous search direction. If this is the case, the line minimization is improved by further corrector steps using a variant of Brent\\'s algorithm.[2]\\nTo summarize: In the first ionic step the forces are calculated for the initial configuration read from the POSCAR file, the second step is a trial (or predictor step), the third step is a corrector step. If the line minimization is sufficiently accurate in this step, the next trial step is performed.\\n\\nNSTEP:\\ninitial positions\\ntrial step\\ncorrector step, i.e. positions corresponding to anticipated minimum\\ntrial step\\ncorrector step\\n...\\nIBRION=3: ionic relaxation (damped molecular dynamics).\\nIf a damping factor is supplied in the INCAR file by means of the SMASS tag, a damped second order equation of motion is used for the update of the ionic degrees of freedom:\\n\\nx\\n→\\n\\n¨\\n\\n=\\n−\\n2\\nα\\n\\nF\\n→\\n\\n−\\nμ\\n\\nx\\n→\\n\\n˙\\n\\n,\\n\\n{\\\\displaystyle {\\\\ddot {\\\\vec {x}}}=-2\\\\alpha {\\\\vec {F}}-\\\\mu {\\\\dot {\\\\vec {x}}},}\\n\\nwhere SMASS supplies the damping factor μ, and POTIM controls α. A simple velocity Verlet algorithm is used to integrate the equation, the discretised equation reads: \\n\\nv\\n→\\n\\nN\\n+\\n1\\n\\n/\\n\\n2\\n\\n=\\n\\n(\\n\\n(\\n1\\n−\\nμ\\n\\n/\\n\\n2\\n)\\n\\nv\\n→\\n\\nN\\n−\\n1\\n\\n/\\n\\n2\\n\\n−\\n2\\nα\\n\\nF\\n→\\n\\nN\\n\\n)\\n\\n/\\n\\n(\\n1\\n+\\nμ\\n\\n/\\n\\n2\\n)\\n\\nx\\n→\\n\\nN\\n+\\n1\\n\\n=\\n\\nx\\n→\\n\\nN\\n+\\n1\\n\\n+\\n\\nv\\n→\\n\\nN\\n+\\n1\\n\\n/\\n\\n2\\n\\n{\\\\displaystyle {\\\\begin{aligned}{{\\\\vec {v}}_{N+1/2}}=&{\\\\Big (}(1-\\\\mu /2){{\\\\vec {v}}_{N-1/2}}-2\\\\alpha {{\\\\vec {F}}_{N}}{\\\\Big )}/(1+\\\\mu /2)\\\\\\\\{{\\\\vec {x}}_{N+1}}=&{{\\\\vec {x}}_{N+1}}+{{\\\\vec {v}}_{N+1/2}}\\\\end{aligned}}}\\n\\nOne may immediately recognize, that μ=2 is equivalent to a simple steepest descent algorithm (of course without line optimization). Hence, μ=2 corresponds to maximal damping, μ=0 corresponds to no damping. The optimal damping factor depends on the Hessian matrix (matrix of the second derivatives of the energy with respect to the atomic positions). A reasonable first guess for μ is usually 0.4.\\nMind that our implementation is particular user-friendly, since changing μ usually does not require to re-adjust the time step POTIM. To choose an optimal time step and damping factor, we recommend the following two step procedure: First fix μ (for instance to 1) and adjust POTIM. POTIM should be chosen as large as possible without getting divergence in the total energy. Then decrease μ and keep POTIM fixed. If POTIM and SMASS are chosen correctly, the damped molecular dynamics mode usually outperforms the conjugate gradient method by a factor of two.\\nIf SMASS is not set in the INCAR file (respectively SMASS<0), a velocity quench algorithm is used. In this case the ionic positions are updated according using the following algorithm: F are the current forces, and α equals POTIM. This equation implies that, if the forces are antiparallel to the velocities, the velocities are quenched to zero. Otherwise the velocities are made parallel to the present forces, and they are increased by an amount that is proportional to the forces.\\nMind: For IBRION=3, a reasonable time step must be supplied by the POTIM parameter. Too large time steps will result in divergence, too small ones will slow down the convergence. The stable time step is usually twice the smallest line minimization step in the conjugate gradient algorithm.\\n\\nIBRION=5 and 6: second derivatives, Hessian matrix, and phonon frequencies (finite differences).\\nWhen IBRION=5 or IBRION=6 are set VASP computes the second-order derivatives of the total energy with respect to the position of the ions using a finite differences approach,\\nthe dynamical matrix is constructed and diagonalized and the phonon modes and frequencies of the system are reported in the OUTCAR file. For precise results, it is necessary to set EDIFF to values less or equal 1E-6. Using the default (EDIFF=1E-4) often results in unacceptably large errors.\\nIBRION=5 does not apply symmetry, whereas IBRION=6 uses symmetry to reduce the number of displacements.\\nIf IBRION=6 and ISIF>=3 the elastic tensors and internal strain tensors are computed as well.\\nBorn effective charges, piezoelectric constants, and the ionic contribution to the dielectric tensor can be\\ncalculated additionally by specifying LEPSILON=.TRUE. (linear response theory) or LCALCEPS=.TRUE. (finite external field).\\n\\nIBRION=7 and 8: second derivatives, Hessian matrix, and phonon frequencies (perturbation theory).\\nWhen IBRION=7 and IBRION=8 are set VASP computes second derivatives of the total energy with respect to the position of the ions using density-functional-perturbation theory (DFPT),\\nthe dynamical matrix is constructed and diagonalized and the phonon modes and frequencies of the system are reported in the OUTCAR file. For precise results, it is necessary to set EDIFF to values around 1E-6. Using the default (EDIFF=1E-4) can result in substantial errors for softer modes.\\nIBRION=7 does not apply symmetry, whereas IBRION=8 uses symmetry to reduce the number of displacements.\\nThe output is similar for IBRION=5 and 6 with the exception that it does not determine the elastic tensors, since the linear response with respect to the strain tensor is not implemented.\\n\\nIBRION=40: calculation of energy profile along IRC.\\nThis method is described in the IRC calculations section.\\n\\nIBRION=44: the Improved Dimer Method.\\nThis method is described in the Improved Dimer Method section.\\n\\nSome general comments\\nFor IBRION=1, 2, and 3, the flag ISIF determines whether the ions and/or the cell shape is changed. Update of the cell shape is supported for molecular dynamics (IBRION=0) only if the dynamics module of Tomas Bucko (precompiler flag -Dtbdyn) is used.\\nWithin all relaxation algorithms (IBRION=1, 2, and 3) the parameter POTIM should be supplied in the INCAR file. For IBRION>0, the forces are scaled internally before calling the minimization routine. Therefore for relaxations, POTIM has no physical meaning and serves only as a scaling factor. For many systems, the optimal POTIM is around 0.5. Because the Quasi-Newton algorithm and the damped algorithms are sensitive to the choice of this parameter, use IBRION=2, if you are not sure how large the optimal POTIM is.\\nIn this case, the OUTCAR file and stdout will contain a line indicating a reliable POTIM. For IBRION=2, the following lines will be written to stdout after each corrector step (usually each odd step):\\n\\n trial: gam=  .00000 g(F)=   .152E+01 g(S)=  .000E+00 ort = .000E+00\\n (trialstep = .82)\\n\\nThe quantity gam is the conjugation parameter to the previous step, g(F) and g(S) are the norm of the force respectively the norm of the stress tensor. The quantity ort is an indicator whether this search direction is orthogonal to the last search direction (for an optimal step this quantity should be much smaller than (g(F) + g(S)). The quantity trialstep is the size of the current trialstep. This value is the average step size leading to a line minimization in the previous ionic step. An optimal POTIM can be determined, by multiplying the current POTIM with the quantity trialstep.\\nAfter at the end of a trial step, the following lines are written to stdout:\\n\\n trial-energy change:   -1.153185  1.order   -1.133   -1.527  -.739\\n step:   1.7275(harm=  2.0557)  dis=  .12277  \\n                            next Energy= -1341.57 (dE= -.142E+01)\\n\\nThe quantity trial-energy change is the change of the energy in the trial step. The first value after 1.order is the expected energy change calculated from the forces: (F(start)+F(trial))/2×change of positions. The second and third value corresponds to  F(start)×change of positions, and F(trial)×change of positions.\\nThe first value in the second line is the size of the step leading to a line minimization along the current search direction. It is calculated from a third order interpolation formula using data from the start and trial step (forces and energy change). harm is the optimal step using a second order (or harmonic) interpolation. Only information on the forces is used for the harmonic interpolation. Close to the minimum both values should be similar. dis is the maximum distance moved by the ions in fractional (direct) coordinates. next Energy gives an indication how large the next energy should be (i.e. the energy at the minimum of the line minimization), dE is the estimated energy change.\\nThe OUTCAR file will contain the following lines, at the end of each trial step:\\n\\n trial-energy change:   -1.148928  1.order   -1.126  -1.518  -.735\\n  (g-gl).g =  .152E+01      g.g   =  .152E+01  gl.gl    =  .000E+00\\n g(Force)  =  .152E+01   g(Stress)=  .000E+00 ortho     =  .000E+00\\n gamma     =    .00000\\n opt step  =   1.72745  (harmonic =   2.05575) max dist = .12277085\\n next E    = -1341.577507   (d E  =   1.42496)\\n\\nThe line trial-energy change was already discussed. g(Force) corresponds to g(F), g(Stress) to g(S), ortho to ort, gamma to gam. The values after gamma correspond to the second line (step: ...) previously described.\\n\\nRelated Tags and Sections\\nNSW,\\nSMASS,\\nPOTIM,\\nNFREE,\\nISIF,\\nLEPSILON,\\nLCALCEPS,\\nImproved Dimer Method\\nExamples that use this tag\\n\\nReferences\\n\\n↑ P. Pulay, Chem. Phys. Lett. 73, 393 (1980).\\n\\n↑ a b W. H. Press, B. P. Flannery, S. A. Teukolsky, and W. T. Vetterling, em Numerical Recipes (Cambridge University Press, New York, 1986).\\n\\nRetrieved from \"https://www.vasp.at/wiki/index.php?title=IBRION&oldid=23897\"'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 静态方法\n",
    "VaspDoc.get_help(\"IBRION\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/leubuntu/src/miniconda3/envs/base_ysl/lib/python3.11/site-packages/urllib3/connectionpool.py:1103: InsecureRequestWarning: Unverified HTTPS request is being made to host '127.0.0.1'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IBRION = -1 | 0 | 1 | 2 | 3 | 5 | 6 | 7 | 8 | 40 | 44 \n",
      "\n",
      "Default: IBRION\n",
      "\n",
      "= -1\n",
      "\n",
      "for NSW=−1 or 0\n",
      "\n",
      "= 0\n",
      "\n",
      "else\n",
      "\n",
      "Description: IBRION determines how the ions are updated and moved.\n",
      "\n",
      "For IBRION=0, a molecular dynamics is performed, whereas all other algorithms are destined for relaxations into a local energy minimum. For difficult relaxation problems it is recommended to use the conjugate gradient algorithm (IBRION=2), which presently possesses the most reliable backup routines. Damped molecular dynamics (IBRION=3) are often useful when starting from very bad initial guesses. Close to the local minimum the RMM-DIIS (IBRION=1) is usually the best choice. IBRION=5 and IBRION=6 are using finite differences to determine the second derivatives (Hessian matrix and phonon frequencies), whereas IBRION=7 and IBRION=8 use density functional perturbation theory to calculate the derivatives.\n",
      "\n",
      "IBRION=-1: no update.\n",
      "The ions are not moved, but NSW outer loops are performed. In each outer loop the electronic degrees of freedom are re-optimized (for NSW>0 this obviously does not make much sense, except for test purposes). If no ionic update is required use NSW=0 instead.\n",
      "\n",
      "IBRION=0: molecular dynamics.\n",
      "Standard ab-initio molecular dynamics. A Verlet algorithm (or fourth-order predictor-corrector if VASP was linked with stepprecor.o) is used to integrate Newton's equations of motion. POTIM supplies the timestep in femto seconds. The parameter SMASS provides additional control.\n",
      "\n",
      "Tip: NWRITE=0,1\n",
      "\n",
      "IBRION=1: ionic relaxation (RMM-DIIS).\n",
      "For IBRION=1, a quasi-Newton (variable metric) algorithm is used to relax the ions into their instantaneous groundstate. The forces and the stress tensor are used to determine the search directions for finding the equilibrium positions (the total energy is not taken into account). This algorithm is very fast and efficient close to local minima, but fails badly if the initial positions are a bad guess (use IBRION=2 in that case). Since the algorithm builds up an approximation of the Hessian matrix it requires very accurate forces, otherwise it will fail to converge. An efficient way to achieve this is to set NELMIN to a value between 4 and 8 (for simple bulk materials 4 is usually adequate, whereas 8 might be required for complex surfaces where the charge density converges very slowly). This forces a minimum of 4 to 8 electronic steps between each ionic step, and guarantees that the forces are well converged at each step.\n",
      "The implemented algorithm is called RMM-DIIS.[1] It implicitly calculates an approximation of the inverse Hessian matrix by taking into account information from previous iterations. On startup, the initial Hessian matrix is diagonal and equal to POTIM. Information from old steps (which can lead to linear dependencies) is automatically removed from the iteration history, if required. The number of vectors kept in the iterations history (which corresponds to the rank of the Hessian matrix must not exceed the degrees of freedom. Naively the number of degrees of freedom is 3(NIONS-1). But symmetry arguments or constraints can reduce this number significantly.\n",
      "There are two algorithms build in to remove information from the iteration history:\n",
      "\n",
      "If NFREE is set in the INCAR file, only up to NFREE ionic steps are kept in the iteration history (the rank of the approximate Hessian matrix is not larger than NFREE).\n",
      "If NFREE is not specified, the criterion whether information is removed from the iteration history is based on the eigenvalue spectrum of the inverse Hessian matrix: if one eigenvalue of the inverse Hessian matrix is larger than 8, information from previous steps is discarded.\n",
      "For complex problems NFREE can usually be set to a rather large value (i.e. 10-20), however systems of low dimensionality require a carful setting of NFREE (or preferably an exact counting of the number of degrees of freedom). To increase NFREE beyond 20 rarely improves convergence. If NFREE is set to too large, the RMM-DIIS algorithm might diverge.\n",
      "The choice of a reasonable POTIM is also important and can speed up calculations significantly, we recommend to find an optimal POTIM using IBRION=2 or performing a few test calculations (see below).\n",
      "\n",
      "IBRION=2: ionic relaxation (conjugate gradient algorithm).\n",
      "A conjugate-gradient algorithm (a discussion of this algorithm can be found for instance in Numerical Recipes, by Press et al.[2]) is used to relax the ions into their instantaneous groundstate. In the first step ions (and cell shape) are changed along the direction of the steepest descent (i.e. the direction of the calculated forces and stress tensor). The conjugate gradient method requires a line minimization, which is performed in several steps: \n",
      "\n",
      "First a trial step into the search direction (scaled gradients) is done, with the length of the trial step controlled by the POTIM tag. Then the energy and the forces are recalculated.\n",
      "The approximate minimum of the total energy is calculated from a cubic (or quadratic) interpolation taking into account the change of the total energy and the change of the forces (3 pieces of information), then a corrector step to the approximate minimum is performed.\n",
      "After the corrector step the forces and energy are recalculated and it is checked whether the forces contain a significant component parallel to the previous search direction. If this is the case, the line minimization is improved by further corrector steps using a variant of Brent's algorithm.[2]\n",
      "To summarize: In the first ionic step the forces are calculated for the initial configuration read from the POSCAR file, the second step is a trial (or predictor step), the third step is a corrector step. If the line minimization is sufficiently accurate in this step, the next trial step is performed.\n",
      "\n",
      "NSTEP:\n",
      "initial positions\n",
      "trial step\n",
      "corrector step, i.e. positions corresponding to anticipated minimum\n",
      "trial step\n",
      "corrector step\n",
      "...\n",
      "IBRION=3: ionic relaxation (damped molecular dynamics).\n",
      "If a damping factor is supplied in the INCAR file by means of the SMASS tag, a damped second order equation of motion is used for the update of the ionic degrees of freedom:\n",
      "\n",
      "x\n",
      "→\n",
      "\n",
      "¨\n",
      "\n",
      "=\n",
      "−\n",
      "2\n",
      "α\n",
      "\n",
      "F\n",
      "→\n",
      "\n",
      "−\n",
      "μ\n",
      "\n",
      "x\n",
      "→\n",
      "\n",
      "˙\n",
      "\n",
      ",\n",
      "\n",
      "{\\displaystyle {\\ddot {\\vec {x}}}=-2\\alpha {\\vec {F}}-\\mu {\\dot {\\vec {x}}},}\n",
      "\n",
      "where SMASS supplies the damping factor μ, and POTIM controls α. A simple velocity Verlet algorithm is used to integrate the equation, the discretised equation reads: \n",
      "\n",
      "v\n",
      "→\n",
      "\n",
      "N\n",
      "+\n",
      "1\n",
      "\n",
      "/\n",
      "\n",
      "2\n",
      "\n",
      "=\n",
      "\n",
      "(\n",
      "\n",
      "(\n",
      "1\n",
      "−\n",
      "μ\n",
      "\n",
      "/\n",
      "\n",
      "2\n",
      ")\n",
      "\n",
      "v\n",
      "→\n",
      "\n",
      "N\n",
      "−\n",
      "1\n",
      "\n",
      "/\n",
      "\n",
      "2\n",
      "\n",
      "−\n",
      "2\n",
      "α\n",
      "\n",
      "F\n",
      "→\n",
      "\n",
      "N\n",
      "\n",
      ")\n",
      "\n",
      "/\n",
      "\n",
      "(\n",
      "1\n",
      "+\n",
      "μ\n",
      "\n",
      "/\n",
      "\n",
      "2\n",
      ")\n",
      "\n",
      "x\n",
      "→\n",
      "\n",
      "N\n",
      "+\n",
      "1\n",
      "\n",
      "=\n",
      "\n",
      "x\n",
      "→\n",
      "\n",
      "N\n",
      "+\n",
      "1\n",
      "\n",
      "+\n",
      "\n",
      "v\n",
      "→\n",
      "\n",
      "N\n",
      "+\n",
      "1\n",
      "\n",
      "/\n",
      "\n",
      "2\n",
      "\n",
      "{\\displaystyle {\\begin{aligned}{{\\vec {v}}_{N+1/2}}=&{\\Big (}(1-\\mu /2){{\\vec {v}}_{N-1/2}}-2\\alpha {{\\vec {F}}_{N}}{\\Big )}/(1+\\mu /2)\\\\{{\\vec {x}}_{N+1}}=&{{\\vec {x}}_{N+1}}+{{\\vec {v}}_{N+1/2}}\\end{aligned}}}\n",
      "\n",
      "One may immediately recognize, that μ=2 is equivalent to a simple steepest descent algorithm (of course without line optimization). Hence, μ=2 corresponds to maximal damping, μ=0 corresponds to no damping. The optimal damping factor depends on the Hessian matrix (matrix of the second derivatives of the energy with respect to the atomic positions). A reasonable first guess for μ is usually 0.4.\n",
      "Mind that our implementation is particular user-friendly, since changing μ usually does not require to re-adjust the time step POTIM. To choose an optimal time step and damping factor, we recommend the following two step procedure: First fix μ (for instance to 1) and adjust POTIM. POTIM should be chosen as large as possible without getting divergence in the total energy. Then decrease μ and keep POTIM fixed. If POTIM and SMASS are chosen correctly, the damped molecular dynamics mode usually outperforms the conjugate gradient method by a factor of two.\n",
      "If SMASS is not set in the INCAR file (respectively SMASS<0), a velocity quench algorithm is used. In this case the ionic positions are updated according using the following algorithm: F are the current forces, and α equals POTIM. This equation implies that, if the forces are antiparallel to the velocities, the velocities are quenched to zero. Otherwise the velocities are made parallel to the present forces, and they are increased by an amount that is proportional to the forces.\n",
      "Mind: For IBRION=3, a reasonable time step must be supplied by the POTIM parameter. Too large time steps will result in divergence, too small ones will slow down the convergence. The stable time step is usually twice the smallest line minimization step in the conjugate gradient algorithm.\n",
      "\n",
      "IBRION=5 and 6: second derivatives, Hessian matrix, and phonon frequencies (finite differences).\n",
      "When IBRION=5 or IBRION=6 are set VASP computes the second-order derivatives of the total energy with respect to the position of the ions using a finite differences approach,\n",
      "the dynamical matrix is constructed and diagonalized and the phonon modes and frequencies of the system are reported in the OUTCAR file. For precise results, it is necessary to set EDIFF to values less or equal 1E-6. Using the default (EDIFF=1E-4) often results in unacceptably large errors.\n",
      "IBRION=5 does not apply symmetry, whereas IBRION=6 uses symmetry to reduce the number of displacements.\n",
      "If IBRION=6 and ISIF>=3 the elastic tensors and internal strain tensors are computed as well.\n",
      "Born effective charges, piezoelectric constants, and the ionic contribution to the dielectric tensor can be\n",
      "calculated additionally by specifying LEPSILON=.TRUE. (linear response theory) or LCALCEPS=.TRUE. (finite external field).\n",
      "\n",
      "IBRION=7 and 8: second derivatives, Hessian matrix, and phonon frequencies (perturbation theory).\n",
      "When IBRION=7 and IBRION=8 are set VASP computes second derivatives of the total energy with respect to the position of the ions using density-functional-perturbation theory (DFPT),\n",
      "the dynamical matrix is constructed and diagonalized and the phonon modes and frequencies of the system are reported in the OUTCAR file. For precise results, it is necessary to set EDIFF to values around 1E-6. Using the default (EDIFF=1E-4) can result in substantial errors for softer modes.\n",
      "IBRION=7 does not apply symmetry, whereas IBRION=8 uses symmetry to reduce the number of displacements.\n",
      "The output is similar for IBRION=5 and 6 with the exception that it does not determine the elastic tensors, since the linear response with respect to the strain tensor is not implemented.\n",
      "\n",
      "IBRION=40: calculation of energy profile along IRC.\n",
      "This method is described in the IRC calculations section.\n",
      "\n",
      "IBRION=44: the Improved Dimer Method.\n",
      "This method is described in the Improved Dimer Method section.\n",
      "\n",
      "Some general comments\n",
      "For IBRION=1, 2, and 3, the flag ISIF determines whether the ions and/or the cell shape is changed. Update of the cell shape is supported for molecular dynamics (IBRION=0) only if the dynamics module of Tomas Bucko (precompiler flag -Dtbdyn) is used.\n",
      "Within all relaxation algorithms (IBRION=1, 2, and 3) the parameter POTIM should be supplied in the INCAR file. For IBRION>0, the forces are scaled internally before calling the minimization routine. Therefore for relaxations, POTIM has no physical meaning and serves only as a scaling factor. For many systems, the optimal POTIM is around 0.5. Because the Quasi-Newton algorithm and the damped algorithms are sensitive to the choice of this parameter, use IBRION=2, if you are not sure how large the optimal POTIM is.\n",
      "In this case, the OUTCAR file and stdout will contain a line indicating a reliable POTIM. For IBRION=2, the following lines will be written to stdout after each corrector step (usually each odd step):\n",
      "\n",
      " trial: gam=  .00000 g(F)=   .152E+01 g(S)=  .000E+00 ort = .000E+00\n",
      " (trialstep = .82)\n",
      "\n",
      "The quantity gam is the conjugation parameter to the previous step, g(F) and g(S) are the norm of the force respectively the norm of the stress tensor. The quantity ort is an indicator whether this search direction is orthogonal to the last search direction (for an optimal step this quantity should be much smaller than (g(F) + g(S)). The quantity trialstep is the size of the current trialstep. This value is the average step size leading to a line minimization in the previous ionic step. An optimal POTIM can be determined, by multiplying the current POTIM with the quantity trialstep.\n",
      "After at the end of a trial step, the following lines are written to stdout:\n",
      "\n",
      " trial-energy change:   -1.153185  1.order   -1.133   -1.527  -.739\n",
      " step:   1.7275(harm=  2.0557)  dis=  .12277  \n",
      "                            next Energy= -1341.57 (dE= -.142E+01)\n",
      "\n",
      "The quantity trial-energy change is the change of the energy in the trial step. The first value after 1.order is the expected energy change calculated from the forces: (F(start)+F(trial))/2×change of positions. The second and third value corresponds to  F(start)×change of positions, and F(trial)×change of positions.\n",
      "The first value in the second line is the size of the step leading to a line minimization along the current search direction. It is calculated from a third order interpolation formula using data from the start and trial step (forces and energy change). harm is the optimal step using a second order (or harmonic) interpolation. Only information on the forces is used for the harmonic interpolation. Close to the minimum both values should be similar. dis is the maximum distance moved by the ions in fractional (direct) coordinates. next Energy gives an indication how large the next energy should be (i.e. the energy at the minimum of the line minimization), dE is the estimated energy change.\n",
      "The OUTCAR file will contain the following lines, at the end of each trial step:\n",
      "\n",
      " trial-energy change:   -1.148928  1.order   -1.126  -1.518  -.735\n",
      "  (g-gl).g =  .152E+01      g.g   =  .152E+01  gl.gl    =  .000E+00\n",
      " g(Force)  =  .152E+01   g(Stress)=  .000E+00 ortho     =  .000E+00\n",
      " gamma     =    .00000\n",
      " opt step  =   1.72745  (harmonic =   2.05575) max dist = .12277085\n",
      " next E    = -1341.577507   (d E  =   1.42496)\n",
      "\n",
      "The line trial-energy change was already discussed. g(Force) corresponds to g(F), g(Stress) to g(S), ortho to ort, gamma to gam. The values after gamma correspond to the second line (step: ...) previously described.\n",
      "\n",
      "Related Tags and Sections\n",
      "NSW,\n",
      "SMASS,\n",
      "POTIM,\n",
      "NFREE,\n",
      "ISIF,\n",
      "LEPSILON,\n",
      "LCALCEPS,\n",
      "Improved Dimer Method\n",
      "Examples that use this tag\n",
      "\n",
      "References\n",
      "\n",
      "↑ P. Pulay, Chem. Phys. Lett. 73, 393 (1980).\n",
      "\n",
      "↑ a b W. H. Press, B. P. Flannery, S. A. Teukolsky, and W. T. Vetterling, em Numerical Recipes (Cambridge University Press, New York, 1986).\n",
      "\n",
      "Retrieved from \"https://www.vasp.at/wiki/index.php?title=IBRION&oldid=23897\"\n"
     ]
    }
   ],
   "source": [
    "# 需初始化\n",
    "VaspDoc().print_help(\"IBRION\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/leubuntu/src/miniconda3/envs/base_ysl/lib/python3.11/site-packages/urllib3/connectionpool.py:1103: InsecureRequestWarning: Unverified HTTPS request is being made to host '127.0.0.1'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings\n",
      "  warnings.warn(\n",
      "/home/leubuntu/src/miniconda3/envs/base_ysl/lib/python3.11/site-packages/pymatgen/io/vasp/help.py:51: GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system (\"lxml\"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.\n",
      "\n",
      "The code that caused this warning is on line 51 of the file /home/leubuntu/src/miniconda3/envs/base_ysl/lib/python3.11/site-packages/pymatgen/io/vasp/help.py. To get rid of this warning, pass the additional argument 'features=\"lxml\"' to the BeautifulSoup constructor.\n",
      "\n",
      "  soup = BeautifulSoup(response.text)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div class=\"mw-body-content mw-content-ltr\" dir=\"ltr\" id=\"mw-content-text\" lang=\"en\"><div class=\"mw-parser-output\"><p><a class=\"mw-selflink selflink\">IBRION</a> = -1 | 0 | 1 | 2 | 3 | 5 | 6 | 7 | 8 | 40 | 44 \n",
       "</p>\n",
       "<table>\n",
       "<tbody><tr>\n",
       "<td>Default: <b>IBRION</b>\n",
       "</td>\n",
       "<td>= -1\n",
       "</td>\n",
       "<td>for <a href=\"/wiki/index.php/NSW\" title=\"NSW\">NSW</a>=−1 or 0\n",
       "</td></tr>\n",
       "<tr>\n",
       "<td>\n",
       "</td>\n",
       "<td>= 0\n",
       "</td>\n",
       "<td>else\n",
       "</td></tr>\n",
       "</tbody></table>\n",
       "<p>Description: <a class=\"mw-selflink selflink\">IBRION</a> determines how the ions are updated and moved.\n",
       "</p>\n",
       "<hr/>\n",
       "<p>For <a class=\"mw-selflink selflink\">IBRION</a>=0, a molecular dynamics is performed, whereas all other algorithms are destined for relaxations into a local energy minimum. For difficult relaxation problems it is recommended to use the conjugate gradient algorithm (<a class=\"mw-selflink selflink\">IBRION</a>=2), which presently possesses the most reliable backup routines. Damped molecular dynamics (<a class=\"mw-selflink selflink\">IBRION</a>=3) are often useful when starting from very bad initial guesses. Close to the local minimum the RMM-DIIS (<a class=\"mw-selflink selflink\">IBRION</a>=1) is usually the best choice. <a class=\"mw-selflink selflink\">IBRION</a>=5 and <a class=\"mw-selflink selflink\">IBRION</a>=6 are using finite differences to determine the second derivatives (Hessian matrix and phonon frequencies), whereas <a class=\"mw-selflink selflink\">IBRION</a>=7 and <a class=\"mw-selflink selflink\">IBRION</a>=8 use density functional perturbation theory to calculate the derivatives.\n",
       "</p>\n",
       "<meta property=\"mw:PageProp/toc\"/>\n",
       "<h2><span id=\"IBRION.3D-1:_no_update.\"></span><span class=\"mw-headline\" id=\"IBRION=-1:_no_update.\"><a class=\"mw-selflink selflink\">IBRION</a>=-1: no update.</span></h2>\n",
       "<p>The ions are not moved, but <a href=\"/wiki/index.php/NSW\" title=\"NSW\">NSW</a> outer loops are performed. In each outer loop the electronic degrees of freedom are re-optimized (for <a href=\"/wiki/index.php/NSW\" title=\"NSW\">NSW</a>&gt;0 this obviously does not make much sense, except for test purposes). If no ionic update is required use <a href=\"/wiki/index.php/NSW\" title=\"NSW\">NSW</a>=0 instead.\n",
       "</p>\n",
       "<h2><span id=\"IBRION.3D0:_molecular_dynamics.\"></span><span class=\"mw-headline\" id=\"IBRION=0:_molecular_dynamics.\"><a class=\"mw-selflink selflink\">IBRION</a>=0: molecular dynamics.</span></h2>\n",
       "<p>Standard ab-initio molecular dynamics. A Verlet algorithm (or fourth-order predictor-corrector if VASP was linked with stepprecor.o) is used to integrate Newton's equations of motion. <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> supplies the timestep in femto seconds. The parameter <a href=\"/wiki/index.php/SMASS\" title=\"SMASS\">SMASS</a> provides additional control.\n",
       "</p>\n",
       "<table style=\"border: 0px solid #2C68FC; padding: 5px; background: #F0F4FF\">\n",
       "<tbody><tr>\n",
       "<td><b><span style=\"color: #2C68FC;\">Tip:</span></b> <a href=\"/wiki/index.php/NWRITE\" title=\"NWRITE\">NWRITE</a>=0,1\n",
       "</td></tr></tbody></table>\n",
       "<h2><span id=\"IBRION.3D1:_ionic_relaxation_.28RMM-DIIS.29.\"></span><span class=\"mw-headline\" id=\"IBRION=1:_ionic_relaxation_(RMM-DIIS).\"><a class=\"mw-selflink selflink\">IBRION</a>=1: ionic relaxation (RMM-DIIS).</span></h2>\n",
       "<p>For <a class=\"mw-selflink selflink\">IBRION</a>=1, a quasi-Newton (variable metric) algorithm is used to relax the ions into their instantaneous groundstate. The forces and the stress tensor are used to determine the search directions for finding the equilibrium positions (the total energy is not taken into account). This algorithm is very fast and efficient close to local minima, but fails badly if the initial positions are a bad guess (use <a class=\"mw-selflink selflink\">IBRION</a>=2 in that case). Since the algorithm builds up an approximation of the Hessian matrix it requires very accurate forces, otherwise it will fail to converge. An efficient way to achieve this is to set <a href=\"/wiki/index.php/NELMIN\" title=\"NELMIN\">NELMIN</a> to a value between 4 and 8 (for simple bulk materials 4 is usually adequate, whereas 8 might be required for complex surfaces where the charge density converges very slowly). This forces a minimum of 4 to 8 electronic steps between each ionic step, and guarantees that the forces are well converged at each step.\n",
       "</p><p>The implemented algorithm is called RMM-DIIS.<sup class=\"reference\" id=\"cite_ref-pulay:cpl:1980_1-0\"><a href=\"#cite_note-pulay:cpl:1980-1\">[1]</a></sup> It implicitly calculates an approximation of the inverse Hessian matrix by taking into account information from previous iterations. On startup, the initial Hessian matrix is diagonal and equal to <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a>. Information from old steps (which can lead to linear dependencies) is automatically removed from the iteration history, if required. The number of vectors kept in the iterations history (which corresponds to the rank of the Hessian matrix must not exceed the degrees of freedom. Naively the number of degrees of freedom is 3(NIONS-1). But symmetry arguments or constraints can reduce this number significantly.\n",
       "</p><p>There are two algorithms build in to remove information from the iteration history:\n",
       "</p>\n",
       "<ol><li>If <a href=\"/wiki/index.php/NFREE\" title=\"NFREE\">NFREE</a> is set in the <a href=\"/wiki/index.php/INCAR\" title=\"INCAR\">INCAR</a> file, only up to <a href=\"/wiki/index.php/NFREE\" title=\"NFREE\">NFREE</a> ionic steps are kept in the iteration history (the rank of the approximate Hessian matrix is not larger than <a href=\"/wiki/index.php/NFREE\" title=\"NFREE\">NFREE</a>).</li>\n",
       "<li>If <a href=\"/wiki/index.php/NFREE\" title=\"NFREE\">NFREE</a> is not specified, the criterion whether information is removed from the iteration history is based on the eigenvalue spectrum of the inverse Hessian matrix: if one eigenvalue of the inverse Hessian matrix is larger than 8, information from previous steps is discarded.</li></ol>\n",
       "<p>For complex problems <a href=\"/wiki/index.php/NFREE\" title=\"NFREE\">NFREE</a> can usually be set to a rather large value (i.e. 10-20), however systems of low dimensionality require a carful setting of <a href=\"/wiki/index.php/NFREE\" title=\"NFREE\">NFREE</a> (or preferably an exact counting of the number of degrees of freedom). To increase <a href=\"/wiki/index.php/NFREE\" title=\"NFREE\">NFREE</a> beyond 20 rarely improves convergence. If <a href=\"/wiki/index.php/NFREE\" title=\"NFREE\">NFREE</a> is set to too large, the RMM-DIIS algorithm might diverge.\n",
       "</p><p>The choice of a reasonable <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> is also important and can speed up calculations significantly, we recommend to find an optimal <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> using <a class=\"mw-selflink selflink\">IBRION</a>=2 or performing a few test calculations (see below).\n",
       "</p>\n",
       "<h2><span id=\"IBRION.3D2:_ionic_relaxation_.28conjugate_gradient_algorithm.29.\"></span><span class=\"mw-headline\" id=\"IBRION=2:_ionic_relaxation_(conjugate_gradient_algorithm).\"><a class=\"mw-selflink selflink\">IBRION</a>=2: ionic relaxation (conjugate gradient algorithm).</span></h2>\n",
       "<p>A conjugate-gradient algorithm (a discussion of this algorithm can be found for instance in <i>Numerical Recipes</i>, by Press <i>et al.</i><sup class=\"reference\" id=\"cite_ref-press:book:1986_2-0\"><a href=\"#cite_note-press:book:1986-2\">[2]</a></sup>) is used to relax the ions into their instantaneous groundstate. In the first step ions (and cell shape) are changed along the direction of the steepest descent (i.e. the direction of the calculated forces and stress tensor). The conjugate gradient method requires a line minimization, which is performed in several steps: \n",
       "</p>\n",
       "<ol><li>First a trial step into the search direction (scaled gradients) is done, with the length of the trial step controlled by the <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> tag. Then the energy and the forces are recalculated.</li>\n",
       "<li>The approximate minimum of the total energy is calculated from a cubic (or quadratic) interpolation taking into account the change of the total energy and the change of the forces (3 pieces of information), then a corrector step to the approximate minimum is performed.</li>\n",
       "<li>After the corrector step the forces and energy are recalculated and it is checked whether the forces contain a significant component parallel to the previous search direction. If this is the case, the line minimization is improved by further corrector steps using a variant of Brent's algorithm.<sup class=\"reference\" id=\"cite_ref-press:book:1986_2-1\"><a href=\"#cite_note-press:book:1986-2\">[2]</a></sup></li></ol>\n",
       "<p>To summarize: In the first ionic step the forces are calculated for the initial configuration read from the <a href=\"/wiki/index.php/POSCAR\" title=\"POSCAR\">POSCAR</a> file, the second step is a trial (or predictor step), the third step is a corrector step. If the line minimization is sufficiently accurate in this step, the next trial step is performed.\n",
       "</p>\n",
       "<dl><dd>NSTEP:\n",
       "<ol><li>initial positions</li>\n",
       "<li>trial step</li>\n",
       "<li>corrector step, i.e. positions corresponding to anticipated minimum</li>\n",
       "<li>trial step</li>\n",
       "<li>corrector step</li></ol>\n",
       "<dl><dd>...</dd></dl></dd></dl>\n",
       "<h2><span id=\"IBRION.3D3:_ionic_relaxation_.28damped_molecular_dynamics.29.\"></span><span class=\"mw-headline\" id=\"IBRION=3:_ionic_relaxation_(damped_molecular_dynamics).\"><a class=\"mw-selflink selflink\">IBRION</a>=3: ionic relaxation (damped molecular dynamics).</span></h2>\n",
       "<p>If a damping factor is supplied in the <a href=\"/wiki/index.php/INCAR\" title=\"INCAR\">INCAR</a> file by means of the <a href=\"/wiki/index.php/SMASS\" title=\"SMASS\">SMASS</a> tag, a damped second order equation of motion is used for the update of the ionic degrees of freedom:\n",
       "</p>\n",
       "<dl><dd><span class=\"mwe-math-element\"><span class=\"mwe-math-mathml-inline mwe-math-mathml-a11y\" style=\"display: none;\"><math alttext=\"{\\displaystyle {\\ddot {\\vec {x}}}=-2\\alpha {\\vec {F}}-\\mu {\\dot {\\vec {x}}},}\" xmlns=\"http://www.w3.org/1998/Math/MathML\">\n",
       "<semantics>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mstyle displaystyle=\"true\" scriptlevel=\"0\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mover>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mover>\n",
       "<mi>x</mi>\n",
       "<mo stretchy=\"false\">→<!-- → --></mo>\n",
       "</mover>\n",
       "</mrow>\n",
       "<mo>¨<!-- ¨ --></mo>\n",
       "</mover>\n",
       "</mrow>\n",
       "</mrow>\n",
       "<mo>=</mo>\n",
       "<mo>−<!-- − --></mo>\n",
       "<mn>2</mn>\n",
       "<mi>α<!-- α --></mi>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mover>\n",
       "<mi>F</mi>\n",
       "<mo stretchy=\"false\">→<!-- → --></mo>\n",
       "</mover>\n",
       "</mrow>\n",
       "</mrow>\n",
       "<mo>−<!-- − --></mo>\n",
       "<mi>μ<!-- μ --></mi>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mover>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mover>\n",
       "<mi>x</mi>\n",
       "<mo stretchy=\"false\">→<!-- → --></mo>\n",
       "</mover>\n",
       "</mrow>\n",
       "<mo>˙<!-- ˙ --></mo>\n",
       "</mover>\n",
       "</mrow>\n",
       "</mrow>\n",
       "<mo>,</mo>\n",
       "</mstyle>\n",
       "</mrow>\n",
       "<annotation encoding=\"application/x-tex\">{\\displaystyle {\\ddot {\\vec {x}}}=-2\\alpha {\\vec {F}}-\\mu {\\dot {\\vec {x}}},}</annotation>\n",
       "</semantics>\n",
       "</math></span><img alt=\"{{\\ddot  {{\\vec  x}}}}=-2\\alpha {{\\vec  F}}-\\mu {{\\dot  {{\\vec  x}}}},\" aria-hidden=\"true\" class=\"mwe-math-fallback-image-inline\" src=\"https://wikimedia.org/api/rest_v1/media/math/render/svg/55c740509ad32102ad27186470b1f3f05bceadb3\" style=\"vertical-align: -0.838ex; width:16.876ex; height:3.343ex;\"/></span></dd></dl>\n",
       "<p>where <a href=\"/wiki/index.php/SMASS\" title=\"SMASS\">SMASS</a> supplies the damping factor μ, and <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> controls α. A simple velocity Verlet algorithm is used to integrate the equation, the discretised equation reads: \n",
       "</p>\n",
       "<dl><dd><span class=\"mwe-math-element\"><span class=\"mwe-math-mathml-inline mwe-math-mathml-a11y\" style=\"display: none;\"><math alttext=\"{\\displaystyle {\\begin{aligned}{{\\vec {v}}_{N+1/2}}=&amp;{\\Big (}(1-\\mu /2){{\\vec {v}}_{N-1/2}}-2\\alpha {{\\vec {F}}_{N}}{\\Big )}/(1+\\mu /2)\\\\{{\\vec {x}}_{N+1}}=&amp;{{\\vec {x}}_{N+1}}+{{\\vec {v}}_{N+1/2}}\\end{aligned}}}\" xmlns=\"http://www.w3.org/1998/Math/MathML\">\n",
       "<semantics>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mstyle displaystyle=\"true\" scriptlevel=\"0\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mtable columnalign=\"right left right left right left right left right left right left\" columnspacing=\"0em 2em 0em 2em 0em 2em 0em 2em 0em 2em 0em\" displaystyle=\"true\" rowspacing=\"3pt\">\n",
       "<mtr>\n",
       "<mtd>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<msub>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mover>\n",
       "<mi>v</mi>\n",
       "<mo stretchy=\"false\">→<!-- → --></mo>\n",
       "</mover>\n",
       "</mrow>\n",
       "</mrow>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mi>N</mi>\n",
       "<mo>+</mo>\n",
       "<mn>1</mn>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mo>/</mo>\n",
       "</mrow>\n",
       "<mn>2</mn>\n",
       "</mrow>\n",
       "</msub>\n",
       "</mrow>\n",
       "<mo>=</mo>\n",
       "</mtd>\n",
       "<mtd>\n",
       "<mi></mi>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mo maxsize=\"1.623em\" minsize=\"1.623em\">(</mo>\n",
       "</mrow>\n",
       "</mrow>\n",
       "<mo stretchy=\"false\">(</mo>\n",
       "<mn>1</mn>\n",
       "<mo>−<!-- − --></mo>\n",
       "<mi>μ<!-- μ --></mi>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mo>/</mo>\n",
       "</mrow>\n",
       "<mn>2</mn>\n",
       "<mo stretchy=\"false\">)</mo>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<msub>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mover>\n",
       "<mi>v</mi>\n",
       "<mo stretchy=\"false\">→<!-- → --></mo>\n",
       "</mover>\n",
       "</mrow>\n",
       "</mrow>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mi>N</mi>\n",
       "<mo>−<!-- − --></mo>\n",
       "<mn>1</mn>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mo>/</mo>\n",
       "</mrow>\n",
       "<mn>2</mn>\n",
       "</mrow>\n",
       "</msub>\n",
       "</mrow>\n",
       "<mo>−<!-- − --></mo>\n",
       "<mn>2</mn>\n",
       "<mi>α<!-- α --></mi>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<msub>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mover>\n",
       "<mi>F</mi>\n",
       "<mo stretchy=\"false\">→<!-- → --></mo>\n",
       "</mover>\n",
       "</mrow>\n",
       "</mrow>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mi>N</mi>\n",
       "</mrow>\n",
       "</msub>\n",
       "</mrow>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mo maxsize=\"1.623em\" minsize=\"1.623em\">)</mo>\n",
       "</mrow>\n",
       "</mrow>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mo>/</mo>\n",
       "</mrow>\n",
       "<mo stretchy=\"false\">(</mo>\n",
       "<mn>1</mn>\n",
       "<mo>+</mo>\n",
       "<mi>μ<!-- μ --></mi>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mo>/</mo>\n",
       "</mrow>\n",
       "<mn>2</mn>\n",
       "<mo stretchy=\"false\">)</mo>\n",
       "</mtd>\n",
       "</mtr>\n",
       "<mtr>\n",
       "<mtd>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<msub>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mover>\n",
       "<mi>x</mi>\n",
       "<mo stretchy=\"false\">→<!-- → --></mo>\n",
       "</mover>\n",
       "</mrow>\n",
       "</mrow>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mi>N</mi>\n",
       "<mo>+</mo>\n",
       "<mn>1</mn>\n",
       "</mrow>\n",
       "</msub>\n",
       "</mrow>\n",
       "<mo>=</mo>\n",
       "</mtd>\n",
       "<mtd>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<msub>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mover>\n",
       "<mi>x</mi>\n",
       "<mo stretchy=\"false\">→<!-- → --></mo>\n",
       "</mover>\n",
       "</mrow>\n",
       "</mrow>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mi>N</mi>\n",
       "<mo>+</mo>\n",
       "<mn>1</mn>\n",
       "</mrow>\n",
       "</msub>\n",
       "</mrow>\n",
       "<mo>+</mo>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<msub>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mover>\n",
       "<mi>v</mi>\n",
       "<mo stretchy=\"false\">→<!-- → --></mo>\n",
       "</mover>\n",
       "</mrow>\n",
       "</mrow>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mi>N</mi>\n",
       "<mo>+</mo>\n",
       "<mn>1</mn>\n",
       "<mrow class=\"MJX-TeXAtom-ORD\">\n",
       "<mo>/</mo>\n",
       "</mrow>\n",
       "<mn>2</mn>\n",
       "</mrow>\n",
       "</msub>\n",
       "</mrow>\n",
       "</mtd>\n",
       "</mtr>\n",
       "</mtable>\n",
       "</mrow>\n",
       "</mstyle>\n",
       "</mrow>\n",
       "<annotation encoding=\"application/x-tex\">{\\displaystyle {\\begin{aligned}{{\\vec {v}}_{N+1/2}}=&amp;{\\Big (}(1-\\mu /2){{\\vec {v}}_{N-1/2}}-2\\alpha {{\\vec {F}}_{N}}{\\Big )}/(1+\\mu /2)\\\\{{\\vec {x}}_{N+1}}=&amp;{{\\vec {x}}_{N+1}}+{{\\vec {v}}_{N+1/2}}\\end{aligned}}}</annotation>\n",
       "</semantics>\n",
       "</math></span><img alt=\"{\\begin{aligned}{{\\vec  v}_{{N+1/2}}}=&amp;{\\Big (}(1-\\mu /2){{\\vec  v}_{{N-1/2}}}-2\\alpha {{\\vec  F}_{N}}{\\Big )}/(1+\\mu /2)\\\\{{\\vec  x}_{{N+1}}}=&amp;{{\\vec  x}_{{N+1}}}+{{\\vec  v}_{{N+1/2}}}\\end{aligned}}\" aria-hidden=\"true\" class=\"mwe-math-fallback-image-inline\" src=\"https://wikimedia.org/api/rest_v1/media/math/render/svg/e4ccfef5338f0f7ce73082b2c07b0024f63ba444\" style=\"vertical-align: -3.211ex; margin-bottom: -0.294ex; width:48.396ex; height:8.176ex;\"/></span></dd></dl>\n",
       "<p>One may immediately recognize, that μ=2 is equivalent to a simple steepest descent algorithm (of course without line optimization). Hence, μ=2 corresponds to maximal damping, μ=0 corresponds to no damping. The optimal damping factor depends on the Hessian matrix (matrix of the second derivatives of the energy with respect to the atomic positions). A reasonable first guess for μ is usually 0.4.\n",
       "</p><p>Mind that our implementation is particular user-friendly, since changing μ usually does not require to re-adjust the time step <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a>. To choose an optimal time step and damping factor, we recommend the following two step procedure: First fix μ (for instance to 1) and adjust <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a>. <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> should be chosen as large as possible without getting divergence in the total energy. Then decrease μ and keep <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> fixed. If <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> and <a href=\"/wiki/index.php/SMASS\" title=\"SMASS\">SMASS</a> are chosen correctly, the damped molecular dynamics mode usually outperforms the conjugate gradient method by a factor of two.\n",
       "</p><p>If <a href=\"/wiki/index.php/SMASS\" title=\"SMASS\">SMASS</a> is not set in the <a href=\"/wiki/index.php/INCAR\" title=\"INCAR\">INCAR</a> file (respectively <a href=\"/wiki/index.php/SMASS\" title=\"SMASS\">SMASS</a>&lt;0), a velocity quench algorithm is used. In this case the ionic positions are updated according using the following algorithm: <b>F</b> are the current forces, and α equals <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a>. This equation implies that, if the forces are antiparallel to the velocities, the velocities are quenched to zero. Otherwise the velocities are made parallel to the present forces, and they are increased by an amount that is proportional to the forces.\n",
       "</p><p><b>Mind</b>: For <a class=\"mw-selflink selflink\">IBRION</a>=3, a reasonable time step <i>must</i> be supplied by the <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> parameter. Too large time steps will result in divergence, too small ones will slow down the convergence. The stable time step is usually twice the <i>smallest</i> line minimization step in the conjugate gradient algorithm.\n",
       "</p><p><span id=\"ibrion56\">\n",
       "</span></p>\n",
       "<h2><span id=\"IBRION.3D5_and_6:_second_derivatives.2C_Hessian_matrix.2C_and_phonon_frequencies_.28finite_differences.29.\"></span><span class=\"mw-headline\" id=\"IBRION=5_and_6:_second_derivatives,_Hessian_matrix,_and_phonon_frequencies_(finite_differences).\"><a class=\"mw-selflink selflink\">IBRION</a>=5 and 6: second derivatives, Hessian matrix, and phonon frequencies (finite differences).</span></h2>\n",
       "<p>When <a class=\"mw-selflink selflink\">IBRION</a>=5 or <a class=\"mw-selflink selflink\">IBRION</a>=6 are set VASP computes the second-order derivatives of the total energy with respect to the position of the ions using a <a href=\"/wiki/index.php/Phonons_from_finite_differences\" title=\"Phonons from finite differences\">finite differences approach</a>,\n",
       "the dynamical matrix is constructed and diagonalized and the phonon modes and frequencies of the system are reported in the <a href=\"/wiki/index.php/OUTCAR\" title=\"OUTCAR\">OUTCAR</a> file. For precise results, it is necessary to set <a href=\"/wiki/index.php/EDIFF\" title=\"EDIFF\">EDIFF</a> to values less or equal 1E-6. Using the default (<a href=\"/wiki/index.php/EDIFF\" title=\"EDIFF\">EDIFF</a>=1E-4) often results in unacceptably large errors.\n",
       "</p><p><a class=\"mw-selflink selflink\">IBRION</a>=5 does not apply symmetry, whereas <a class=\"mw-selflink selflink\">IBRION</a>=6 uses symmetry to reduce the number of displacements.\n",
       "</p><p>If <a class=\"mw-selflink selflink\">IBRION</a>=6 and <a href=\"/wiki/index.php/ISIF\" title=\"ISIF\">ISIF</a>&gt;=3 the elastic tensors and internal strain tensors are computed as well.\n",
       "</p><p>Born effective charges, piezoelectric constants, and the ionic contribution to the dielectric tensor can be\n",
       "calculated additionally by specifying <a href=\"/wiki/index.php/LEPSILON\" title=\"LEPSILON\">LEPSILON</a>=.TRUE. (linear response theory) or <a href=\"/wiki/index.php/LCALCEPS\" title=\"LCALCEPS\">LCALCEPS</a>=.TRUE. (finite external field).\n",
       "\n",
       "</p><p><span id=\"ibrion78\">\n",
       "</span></p>\n",
       "<h2><span id=\"IBRION.3D7_and_8:_second_derivatives.2C_Hessian_matrix.2C_and_phonon_frequencies_.28perturbation_theory.29.\"></span><span class=\"mw-headline\" id=\"IBRION=7_and_8:_second_derivatives,_Hessian_matrix,_and_phonon_frequencies_(perturbation_theory).\"><a class=\"mw-selflink selflink\">IBRION</a>=7 and 8: second derivatives, Hessian matrix, and phonon frequencies (perturbation theory).</span></h2>\n",
       "<p>When <a class=\"mw-selflink selflink\">IBRION</a>=7 and <a class=\"mw-selflink selflink\">IBRION</a>=8 are set VASP computes second derivatives of the total energy with respect to the position of the ions using <a href=\"/wiki/index.php/Phonons_from_density-functional-perturbation_theory\" title=\"Phonons from density-functional-perturbation theory\">density-functional-perturbation theory</a> (DFPT),\n",
       "the dynamical matrix is constructed and diagonalized and the phonon modes and frequencies of the system are reported in the <a href=\"/wiki/index.php/OUTCAR\" title=\"OUTCAR\">OUTCAR</a> file. For precise results, it is necessary to set <a href=\"/wiki/index.php/EDIFF\" title=\"EDIFF\">EDIFF</a> to values around 1E-6. Using the default (<a href=\"/wiki/index.php/EDIFF\" title=\"EDIFF\">EDIFF</a>=1E-4) can result in substantial errors for softer modes.\n",
       "</p><p><a class=\"mw-selflink selflink\">IBRION</a>=7 does not apply symmetry, whereas <a class=\"mw-selflink selflink\">IBRION</a>=8 uses symmetry to reduce the number of displacements.\n",
       "</p><p>The output is similar for <a class=\"mw-selflink-fragment\" href=\"#ibrion56\"><b>IBRION</b>=5 and 6</a> with the exception that it does not determine the elastic tensors, since the linear response with respect to the strain tensor is not implemented.\n",
       "\n",
       "</p>\n",
       "<h2><span id=\"IBRION.3D40:_calculation_of_energy_profile_along_IRC.\"></span><span class=\"mw-headline\" id=\"IBRION=40:_calculation_of_energy_profile_along_IRC.\"><a class=\"mw-selflink selflink\">IBRION</a>=40: calculation of energy profile along IRC.</span></h2>\n",
       "<p>This method is described in the <a class=\"mw-redirect\" href=\"/wiki/index.php/IRC_calculations\" title=\"IRC calculations\">IRC calculations</a> section.\n",
       "</p>\n",
       "<h2><span id=\"IBRION.3D44:_the_Improved_Dimer_Method.\"></span><span class=\"mw-headline\" id=\"IBRION=44:_the_Improved_Dimer_Method.\"><a class=\"mw-selflink selflink\">IBRION</a>=44: the Improved Dimer Method.</span></h2>\n",
       "<p>This method is described in the <a class=\"mw-redirect\" href=\"/wiki/index.php/Improved_Dimer_Method\" title=\"Improved Dimer Method\">Improved Dimer Method</a> section.\n",
       "</p>\n",
       "<h2><span class=\"mw-headline\" id=\"Some_general_comments\">Some general comments</span></h2>\n",
       "<p>For <a class=\"mw-selflink selflink\">IBRION</a>=1, 2, and 3, the flag <a href=\"/wiki/index.php/ISIF\" title=\"ISIF\">ISIF</a> determines whether the ions and/or the cell shape is changed. Update of the cell shape is supported for molecular dynamics (<a class=\"mw-selflink selflink\">IBRION</a>=0) only if the dynamics module of Tomas Bucko (precompiler flag <tt>-Dtbdyn</tt>) is used.\n",
       "</p><p>Within all relaxation algorithms (<a class=\"mw-selflink selflink\">IBRION</a>=1, 2, and 3) the parameter <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> should be supplied in the <a href=\"/wiki/index.php/INCAR\" title=\"INCAR\">INCAR</a> file. For <a class=\"mw-selflink selflink\">IBRION</a>&gt;0, <i>the forces are scaled internally before calling the minimization routine</i>. Therefore for relaxations, <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> has no physical meaning and serves only as a scaling factor. For many systems, the optimal <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> is around 0.5. Because the Quasi-Newton algorithm and the damped algorithms are sensitive to the choice of this parameter, use <a class=\"mw-selflink selflink\">IBRION</a>=2, if you are not sure how large the optimal <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> is.\n",
       "</p><p>In this case, the <a href=\"/wiki/index.php/OUTCAR\" title=\"OUTCAR\">OUTCAR</a> file and <tt>stdout</tt> will contain a line indicating a reliable <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a>. For <a class=\"mw-selflink selflink\">IBRION</a>=2, the following lines will be written to <tt>stdout</tt> after each corrector step (usually each odd step):\n",
       "</p>\n",
       "<pre> trial: gam=  .00000 g(F)=   .152E+01 g(S)=  .000E+00 ort = .000E+00\n",
       " (trialstep = .82)\n",
       "</pre>\n",
       "<p>The quantity <tt>gam</tt> is the conjugation parameter to the previous step, <tt>g(F)</tt> and <tt>g(S)</tt> are the norm of the force respectively the norm of the stress tensor. The quantity <tt>ort</tt> is an indicator whether this search direction is orthogonal to the last search direction (for an optimal step this quantity should be much smaller than <tt>(g(F) + g(S))</tt>. The quantity <tt>trialstep</tt> is the size of the current trialstep. This value is the average step size leading to a line minimization in the previous ionic step. An optimal <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> can be determined, by multiplying the current <a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a> with the quantity trialstep.\n",
       "</p><p>After at the end of a trial step, the following lines are written to <tt>stdout</tt>:\n",
       "</p>\n",
       "<pre> trial-energy change:   -1.153185  1.order   -1.133   -1.527  -.739\n",
       " step:   1.7275(harm=  2.0557)  dis=  .12277  \n",
       "                            next Energy= -1341.57 (dE= -.142E+01)\n",
       "</pre>\n",
       "<p>The quantity <tt>trial-energy</tt> change is the change of the energy in the trial step. The first value after <tt>1.order</tt> is the expected energy change calculated from the forces: (<b>F</b>(start)+<b>F</b>(trial))/2×change of positions. The second and third value corresponds to  <b>F</b>(start)×change of positions, and <b>F</b>(trial)×change of positions.\n",
       "</p><p>The first value in the second line is the size of the step leading to a line minimization along the current search direction. It is calculated from a third order interpolation formula using data from the start and trial step (forces and energy change). <tt>harm</tt> is the optimal step using a second order (or harmonic) interpolation. Only information on the forces is used for the harmonic interpolation. Close to the minimum both values should be similar. <tt>dis</tt> is the maximum distance moved by the ions in fractional (direct) coordinates. <tt>next Energy</tt> gives an indication how large the next energy should be (i.e. the energy at the minimum of the line minimization), dE is the estimated energy change.\n",
       "</p><p>The <a href=\"/wiki/index.php/OUTCAR\" title=\"OUTCAR\">OUTCAR</a> file will contain the following lines, at the end of each trial step:\n",
       "</p>\n",
       "<pre> trial-energy change:   -1.148928  1.order   -1.126  -1.518  -.735\n",
       "  (g-gl).g =  .152E+01      g.g   =  .152E+01  gl.gl    =  .000E+00\n",
       " g(Force)  =  .152E+01   g(Stress)=  .000E+00 ortho     =  .000E+00\n",
       " gamma     =    .00000\n",
       " opt step  =   1.72745  (harmonic =   2.05575) max dist = .12277085\n",
       " next E    = -1341.577507   (d E  =   1.42496)\n",
       "</pre>\n",
       "<p>The line <tt>trial-energy change</tt> was already discussed. <tt>g(Force)</tt> corresponds to <tt>g(F)</tt>, <tt>g(Stress)</tt> to <tt>g(S)</tt>, <tt>ortho</tt> to <tt>ort</tt>, <tt>gamma</tt> to <tt>gam</tt>. The values after <tt>gamma</tt> correspond to the second line (<tt>step:</tt> ...) previously described.\n",
       "</p>\n",
       "<h2><span class=\"mw-headline\" id=\"Related_Tags_and_Sections\">Related Tags and Sections</span></h2>\n",
       "<p><a href=\"/wiki/index.php/NSW\" title=\"NSW\">NSW</a>,\n",
       "<a href=\"/wiki/index.php/SMASS\" title=\"SMASS\">SMASS</a>,\n",
       "<a href=\"/wiki/index.php/POTIM\" title=\"POTIM\">POTIM</a>,\n",
       "<a href=\"/wiki/index.php/NFREE\" title=\"NFREE\">NFREE</a>,\n",
       "<a href=\"/wiki/index.php/ISIF\" title=\"ISIF\">ISIF</a>,\n",
       "<a href=\"/wiki/index.php/LEPSILON\" title=\"LEPSILON\">LEPSILON</a>,\n",
       "<a href=\"/wiki/index.php/LCALCEPS\" title=\"LCALCEPS\">LCALCEPS</a>,\n",
       "<a class=\"mw-redirect\" href=\"/wiki/index.php/Improved_Dimer_Method\" title=\"Improved Dimer Method\">Improved Dimer Method</a>\n",
       "</p><p><a href=\"/wiki/index.php/Special:Search/%22IBRION%22_incategory:%22Examples%22\" title='Special:Search/\"IBRION\" incategory:\"Examples\"'>Examples that use this tag</a>\n",
       "</p>\n",
       "<h2><span class=\"mw-headline\" id=\"References\">References</span></h2>\n",
       "<div class=\"mw-references-wrap\"><ol class=\"references\">\n",
       "<li id=\"cite_note-pulay:cpl:1980-1\"><a href=\"#cite_ref-pulay:cpl:1980_1-0\">↑</a> <span class=\"reference-text\"><a class=\"external text\" href=\"https://doi.org/10.1016/0009-2614(80)80396-4\" rel=\"nofollow\">P. Pulay, Chem. Phys. Lett. <b>73</b>, 393 (1980).</a></span>\n",
       "</li>\n",
       "<li id=\"cite_note-press:book:1986-2\">↑ <sup><a href=\"#cite_ref-press:book:1986_2-0\">a</a></sup> <sup><a href=\"#cite_ref-press:book:1986_2-1\">b</a></sup> <span class=\"reference-text\"><a class=\"external text\" href=\"https://archive.org/details/numericalrecipes00pres\" rel=\"nofollow\">W. H. Press, B. P. Flannery, S. A. Teukolsky, and W. T. Vetterling, em Numerical Recipes (Cambridge University Press, New York, 1986).</a></span>\n",
       "</li>\n",
       "</ol></div>\n",
       "<hr/>\n",
       "<!-- \n",
       "NewPP limit report\n",
       "Cached time: 20240511210648\n",
       "Cache expiry: 86400\n",
       "Reduced expiry: false\n",
       "Complications: [show‐toc, no‐toc‐conversion]\n",
       "CPU time usage: 0.277 seconds\n",
       "Real time usage: 0.662 seconds\n",
       "Preprocessor visited node count: 14325/1000000\n",
       "Post‐expand include size: 2012/2097152 bytes\n",
       "Template argument size: 43841/2097152 bytes\n",
       "Highest expansion depth: 6/100\n",
       "Expensive parser function count: 0/100\n",
       "Unstrip recursion depth: 0/20\n",
       "Unstrip post‐expand size: 1194/5000000 bytes\n",
       "-->\n",
       "<!--\n",
       "Transclusion expansion time report (%,ms,calls,template)\n",
       "100.00%  168.364      1 -total\n",
       " 81.66%  137.487      3 Template:Cite\n",
       " 47.04%   79.203    966 Template:Reference\n",
       "  4.80%    8.077      1 Template:DEF\n",
       "  2.83%    4.759     95 Template:TAG\n",
       "  2.82%    4.754      9 Template:FILE\n",
       "  2.14%    3.600      1 Template:TAGDEF\n",
       "  2.02%    3.409      1 Template:Sc\n",
       "  1.90%    3.193      1 Template:NB\n",
       "  1.40%    2.352      2 Template:!-\n",
       "-->\n",
       "<!-- Saved in parser cache with key mediawiki-vw:pcache:idhash:167-0!canonical and timestamp 20240511210647 and revision id 23897. Rendering was triggered because: page-view\n",
       " -->\n",
       "</div>\n",
       "<div class=\"printfooter\" data-nosnippet=\"\">Retrieved from \"<a dir=\"ltr\" href=\"https://www.vasp.at/wiki/index.php?title=IBRION&amp;oldid=23897\">https://www.vasp.at/wiki/index.php?title=IBRION&amp;oldid=23897</a>\"</div></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# 需初始化\n",
    "# 展示 HTML 格式内容\n",
    "VaspDoc().print_jupyter_help(\"IBRION\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base_ysl",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
